### PR TITLE
Invalidation

### PR DESCRIPTION
**Pull request by tpietzsch**
Adds `invalidate(K key)`, `invalidateIf(Predicate<K> condition)`, and `invalidateAll()` methods.

*"Invalidate"* means: Remove from the cache, and discard any changes that have been made in this cache or any backing cache. In particular, if changes have been persisted to disk, remove those as well.

### CODE REVIEW

File: src/main/java/net/imglib2/cache/Invalidate.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	 * See {@link AbstractCache#invalidateIf(long, Predicate)},
+	 * {@link CacheRemover#invalidateIf(long, Predicate)}.
+	 *
+	 * @param condition
+	 *            condition on keys of entries to remove
+	 * @param parallelismThreshold
+	 *            the (estimated) number of entries in the cache needed for this
+	 *            operation to be executed in parallel
+	 */
+	void invalidateIf( final long parallelismThreshold, final Predicate< K > condition );
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/Invalidate.java:48]: What about
```java
void invalidateIf( final Predicate< K > condition );
```
in accordance to
```java
void invalidateAll( final long parallelismThreshold );
```

File: src/main/java/net/imglib2/cache/Invalidate.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	/**
+	 * Removes and discards the entry with the specified {@code key}.
+	 * <p>
+	 * See {@link AbstractCache#invalidate(Object)},
+	 * {@link CacheRemover#invalidate(Object)}.
+	 *
+	 * @param key
+	 *            key of the entry to remove
+	 */
+	void invalidate( final K key );
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/Invalidate.java:34]: For convenience, might add
```java
default void invalidate( final K... keys ) {
    invalidate( Arrays.asList( keys ) );
}

default void invalidate( final Collection< ? extends K > keys ) {
    invalidate( Arrays.asList( keys ) );
    keys.forEach( this::invalidate );
}
```

File: src/main/java/net/imglib2/cache/Invalidate.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	 */
+	default void invalidateAll()
+	{
+		invalidateAll( defaultParallelismThreshold );
+	}
+
+	/**
+	 * TODO: Should do some benchmarks. This is picked completely arbitrarily.
+	 */
+	long defaultParallelismThreshold = 5000;
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/Invalidate.java:87]: Not sure I understand the parallelism threshold. Is this the task size (in counts of keys) for parallel jobs if applicable? Should `defaultParallelismthreshold` be configurable?

File: src/main/java/net/imglib2/cache/PausableQueue.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	{
+		final E x = elements.poll();
+		notFull.signal();
+		return x;
+	}
+
+	/**
+	 * Remove all occurrences of {@code element}.
+	 * <p>
+	 * Call only when queue is paused.
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/PausableQueue.java:166]: Could throw an appropriate exception if queue is not paused. Would certainly help debugging.

File: src/main/java/net/imglib2/cache/PausableQueue.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	public void remove( final E element )
+	{
+		final Iterator< E > it = elements.iterator();
+		while ( it.hasNext() )
+			if ( it.next() == element )
+				it.remove();
+	}
+
+	/**
+	 * Remove all elements.
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/PausableQueue.java:177]: Could throw an appropriate exception if queue is not paused. Would certainly help debugging.

File: src/main/java/net/imglib2/cache/PausableQueue.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+	 * <p>
+	 * Call only when queue is paused.
+	 */
+	public void clear()
+	{
+		elements.clear();
+	}
+
+	/**
+	 * Remove all elements matching {@code condition}.
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/PausableQueue.java:187]: Could throw an appropriate exception if queue is not paused. Would certainly help debugging.

File: src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderCache.java [hunk 7]
```diff
@@ -154,10 +155,50 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader )
...
+			if ( ref != null )
+				ref.clear();
+			entry.ref = null;
+		}
+	}
+
+	@Override
+	public void invalidateIf( final long parallelismThreshold, final Predicate< K > condition )
+	{
+		map.forEachValue( parallelismThreshold, entry ->
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderCache.java:174]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderCache.java [hunk 8]
```diff
@@ -154,10 +155,50 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader )
...
+	}
+
+	@Override
+	public void invalidateAll( final long parallelismThreshold )
+	{
+		// TODO: We could also simply do map.clear(). Pros/Cons?
+
+		map.forEachValue( parallelismThreshold, entry ->
+		{
+			entry.remove();
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderCache.java:195]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderRemoverCache.java [hunk 9]
```diff
@@ -204,10 +205,54 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader, f
...
+	public void invalidateIf( final long parallelismThreshold, final Predicate< K > condition )
+	{
+		map.forEachValue( parallelismThreshold, entry ->
+		{
+			if ( condition.test( entry.key ) )
+			{
+				strongCache.invalidate( entry.key );
+				synchronized ( entry )
+				{
+					map.remove( entry.key, entry );
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderRemoverCache.java:233]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderRemoverCache.java [hunk 10]
```diff
@@ -204,10 +205,54 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader, f
...
+	@Override
+	public void invalidateAll( final long parallelismThreshold )
 	{
-		// TODO
-		throw new UnsupportedOperationException( "not implemented yet" );
+		map.forEachValue( parallelismThreshold, entry ->
+		{
+			synchronized ( entry )
+			{
+				map.remove( entry.key, entry );
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderRemoverCache.java:249]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/SoftRefLoaderCache.java [hunk 11]
```diff
@@ -117,9 +118,47 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader )
...
+	}
+
+	@Override
+	public void invalidateIf( final long parallelismThreshold, final Predicate< K > condition )
+	{
+		map.forEachValue( parallelismThreshold, entry ->
+		{
+			if ( condition.test( entry.key ) )
+			{
+				entry.remove();
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/SoftRefLoaderCache.java:140]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/SoftRefLoaderCache.java [hunk 12]
```diff
@@ -117,9 +118,47 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader )
...
+	}
+
+	@Override
+	public void invalidateAll( final long parallelismThreshold )
+	{
+		// TODO: We could also simply do map.clear(). Pros/Cons?
+
+		map.forEachValue( parallelismThreshold, entry ->
+		{
+			entry.remove();
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/SoftRefLoaderCache.java:156]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/SoftRefLoaderRemoverCache.java [hunk 13]
```diff
@@ -178,10 +173,51 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader, f
...
+	@Override
+	public void invalidateIf( final long parallelismThreshold, final Predicate< K > condition )
+	{
+		map.forEachValue( parallelismThreshold, entry ->
+		{
+			if ( condition.test( entry.key ) )
+			{
+				synchronized ( entry )
+				{
+					map.remove( entry.key, entry );
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/SoftRefLoaderRemoverCache.java:199]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/SoftRefLoaderRemoverCache.java [hunk 14]
```diff
@@ -178,10 +173,51 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader, f
...
+	}
+
+	@Override
+	public void invalidateAll( final long parallelismThreshold )
+	{
+		map.forEachValue( parallelismThreshold, entry ->
+		{
+			synchronized ( entry )
+			{
+				map.remove( entry.key, entry );
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/SoftRefLoaderRemoverCache.java:215]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/WeakRefLoaderCache.java [hunk 15]
```diff
@@ -117,10 +118,47 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader )
...
+	}
+
+	@Override
+	public void invalidateIf( final long parallelismThreshold, final Predicate< K > condition )
+	{
+		map.forEachValue( parallelismThreshold, entry ->
+		{
+			if ( condition.test( entry.key ) )
+			{
+				entry.remove();
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/WeakRefLoaderCache.java:140]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/WeakRefLoaderCache.java [hunk 16]
```diff
@@ -117,10 +118,47 @@ public V get( final K key, final CacheLoader< ? super K, ? extends V > loader )
...
+	}
+
+	@Override
+	public void invalidateAll( final long parallelismThreshold )
+	{
+		// TODO: We could also simply do map.clear(). Pros/Cons?
+
+		map.forEachValue( parallelismThreshold, entry ->
+		{
+			entry.remove();
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/WeakRefLoaderCache.java:156]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/WeakRefVolatileCache.java [hunk 17]
```diff
@@ -196,18 +197,106 @@ public void cleanUp()
...
+		{
+			// stop fetcher threads to avoid concurrent load for key
+			fetchQueue.pause();
+
+			// remove matching entries from this cache
+			map.forEachValue( parallelismThreshold, entry ->
+			{
+				if ( condition.test( entry.key ) )
+				{
+					entry.remove();
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/WeakRefVolatileCache.java:250]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/WeakRefVolatileCache.java [hunk 18]
```diff
@@ -196,18 +197,106 @@ public void cleanUp()
...
+		try
+		{
+			// stop fetcher threads to avoid concurrent load for key
+			fetchQueue.pause();
+
+			// remove all entries from this cache
+			// TODO: We could also simply do map.clear(). Pros/Cons?
+			map.forEachValue( parallelismThreshold, entry ->
+			{
+				entry.remove();
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/WeakRefVolatileCache.java:282]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/WeakRefVolatileLoaderCache.java [hunk 19]
```diff
@@ -196,18 +197,108 @@ public void cleanUp()
...
+		{
+			// stop fetcher threads to avoid concurrent load for key
+			fetchQueue.pause();
+
+			// remove matching entries from this cache
+			map.forEachValue( parallelismThreshold, entry ->
+			{
+				if ( condition.test( entry.key ) )
+				{
+					entry.remove();
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/WeakRefVolatileLoaderCache.java:251]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/ref/WeakRefVolatileLoaderCache.java [hunk 20]
```diff
@@ -196,18 +197,108 @@ public void cleanUp()
...
+	{
+		try
+		{
+			// stop fetcher threads to avoid concurrent load for key
+			fetchQueue.pause();
+
+			// remove all entries from this cache
+			// TODO: We could also simply do map.clear(). Pros/Cons?
+			map.forEachValue( parallelismThreshold, entry -> {
+				entry.remove();
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/ref/WeakRefVolatileLoaderCache.java:283]: Is it save to remove `Entry` from `map` while iterating over it?

File: src/main/java/net/imglib2/cache/Invalidate.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * @param <K>
+ *            key type (identifies entries to remove)
+ *
+ * @author Tobias Pietzsch
+ */
+public interface Invalidate< K >
+{
+	/**
+	 * Removes and discards the entry with the specified {@code key}.
```
Comment (hanslovsky) [inline src/main/java/net/imglib2/cache/Invalidate.java:26]: Should probably add that it will not do anything if key is not present.

**Review by hanslovsky — COMMENTED** [review]
I had a first look at it and it looks pretty good! Can't wait to use the `Invalidate` interface. I left a few comments here and there with suggestions or asking for clarification. You will see that
> Is it save to remove `Entry` from `map` while iterating over it?

will appear quite often. I wanted to make sure that I tagged all occurences in case the answer to that question is *no*. Answering it only once is good enough, of course :smile:

File: src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderCache.java (hunk 7 above)
Comment (tpietzsch) [reply src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderCache.java:174]: yes

File: src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderCache.java (hunk 8 above)
Comment (tpietzsch) [reply src/main/java/net/imglib2/cache/ref/GuardedStrongRefLoaderCache.java:195]: yes

**Pull request merged**